File: examples/kmeans.R
require(SparkR)

# Logistic regression in Spark.
# Note: unlike the example in Scala, a point here is represented as a vector of
# doubles.

parseVector <- function(line) {
  nums = strsplit(line, " ")[[1]]
  sapply(nums, as.double)
}

parseVectors <- function(lines) {
	lines = strsplit(as.character(lines) , " ", fixed = TRUE)
	split(matrix(as.numeric(unlist(lines)), ncol = length(lines[[1]])), 1:length(lines))
}


closestPoint <- function(p, centers) {
  bestIndex <- 0
  closest <- .Machine$double.xmax
  for (i in 1:length(centers)) {
    tempDist <- sum((p - centers[[i]]) ** 2)
    if (tempDist < closest) {
      closest <- tempDist
      bestIndex <- i
    }
  }
  bestIndex
}

# Main program

args <- commandArgs(trailing = TRUE)

if (length(args) != 4) {
  print("Usage: kmeans <master> <file> <K> <convergeDist>")
  q("no")
}

sc <- sparkR.init(args[[1]], "RKMeans")
K <- as.integer(args[[3]])
convergeDist <- as.double(args[[4]])

lines <- textFile(sc, args[[2]])
points <- cache(lapplyPartition(lines, parseVectors))
# kPoints <- take(points, K)
kPoints <- takeSample(points, FALSE, K, 16189L)
tempDist <- 1.0

while (tempDist > convergeDist) {
  closest <- lapply(points,
                    function(p) { list(closestPoint(p, kPoints), list(p, 1)) })

  pointStats <- reduceByKey(closest,
                            function(p1, p2) {
                              list(p1[[1]] + p2[[1]], p1[[2]] + p2[[2]])
                            },
                            2L)

  newPoints <- collect(lapply(pointStats,
                              function(tup) {
                                list(tup[[1]], tup[[2]][[1]] / tup[[2]][[2]])
                              }))

  tempDist <- sum(sapply(newPoints,
                         function(tup) {
                           sum((kPoints[[tup[[1]]]] - tup[[2]]) ** 2)
                         }))

  for (tup in newPoints)
    kPoints[[tup[[1]]]] <- tup[[2]]

  cat("Finished iteration (delta = ", tempDist, ")\n")
}

cat("Final centers:\n")
writeLines(unlist(lapply(kPoints, paste, collapse = " ")))

File: examples/src/main/python/kmeans.py
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

"""
The K-means algorithm written from scratch against PySpark. In practice,
one may prefer to use the KMeans algorithm in ML, as shown in
examples/src/main/python/ml/kmeans_example.py.

This example requires NumPy (http://www.numpy.org/).
"""
import sys
from typing import List

import numpy as np
from pyspark.sql import SparkSession


def parseVector(line: str) -> np.ndarray:
    return np.array([float(x) for x in line.split(' ')])


def closestPoint(p: np.ndarray, centers: List[np.ndarray]) -> int:
    bestIndex = 0
    closest = float("+inf")
    for i in range(len(centers)):
        tempDist = np.sum((p - centers[i]) ** 2)
        if tempDist < closest:
            closest = tempDist
            bestIndex = i
    return bestIndex


if __name__ == "__main__":

    if len(sys.argv) != 4:
        print("Usage: kmeans <file> <k> <convergeDist>", file=sys.stderr)
        sys.exit(-1)

    print("""WARN: This is a naive implementation of KMeans Clustering and is given
       as an example! Please refer to examples/src/main/python/ml/kmeans_example.py for an
       example on how to use ML's KMeans implementation.""", file=sys.stderr)

    spark = SparkSession\
        .builder\
        .appName("PythonKMeans")\
        .getOrCreate()

    lines = spark.read.text(sys.argv[1]).rdd.map(lambda r: r[0])
    data = lines.map(parseVector).cache()
    K = int(sys.argv[2])
    convergeDist = float(sys.argv[3])

    kPoints = data.takeSample(False, K, 1)
    tempDist = 1.0

    while tempDist > convergeDist:
        closest = data.map(
            lambda p: (closestPoint(p, kPoints), (p, 1)))
        pointStats = closest.reduceByKey(
            lambda p1_c1, p2_c2: (p1_c1[0] + p2_c2[0], p1_c1[1] + p2_c2[1]))
        newPoints = pointStats.map(
            lambda st: (st[0], st[1][0] / st[1][1])).collect()

        tempDist = sum(np.sum((kPoints[iK] - p) ** 2) for (iK, p) in newPoints)

        for (iK, p) in newPoints:
            kPoints[iK] = p

    print("Final centers: " + str(kPoints))

    spark.stop()
